File: design-pattern-go/builder/html_builder.go
package builder

import (
	"log"
	"os"
)

type HTMLBuilder struct {
	fileName string
	body     string
}

func NewHTMLBuilder() Builder {
	return &HTMLBuilder{
		fileName: "untitle.html",
		body:     "",
	}
}

func (b *HTMLBuilder) MakeTitle(title string) {
	b.fileName = title + ".html"
	b.body += "<!DOCTYPE html>\n"
	b.body += "<html>\n"
	b.body += "<head><title>"
	b.body += title
	b.body += "</title></head>\n"
	b.body += "<body>\n"
	b.body += "<h1>"
	b.body += title
	b.body += "</h1>\n\n"
}

func (b *HTMLBuilder) MakeString(str string) {
	b.body += "<p>"
	b.body += str
	b.body += "</p>\n\n"
}

func (b *HTMLBuilder) MakeItems(items []string) {
	b.body += "<ul>"
	for _, item := range items {
		b.body += "<li>"
		b.body += item
		b.body += "</li>\n"
	}
	b.body += "\n"
}

func (b *HTMLBuilder) Close() {
	file, err := os.Create(b.fileName)
	if err != nil {
		log.Fatal("couldn't open file")
	}
	defer file.Close()

	b.body += "</body>"
	b.body += "</html>\n"

	byteBody := []byte(b.body)
	_, err = file.Write(byteBody)
	if err != nil {
		log.Fatal("couldn't write body to file")
	}
}

func (b *HTMLBuilder) GetHTMLResult() string {
	return b.fileName
}

File: design-pattern-python/design_pattern_python/builder/html_builder.py
from typing import List

from builder import Builder, StringBuilder


class HTMLBuilder(Builder):
    def __init__(self) -> None:
        self.__filename = "untitled.html"
        self.__sb = StringBuilder()

    def make_title(self, title: str) -> None:
        self.__filename = title + ".html"
        self.__sb.append("<!DOCTYPE html> \n")
        self.__sb.append("<html>\n")
        self.__sb.append("<head><title>")
        self.__sb.append(title)
        self.__sb.append("</title></head>\n")
        self.__sb.append("<body>\n")
        self.__sb.append("<h1>")
        self.__sb.append(title)
        self.__sb.append("</h1>\n\n")

    def make_string(self, string: str) -> None:
        self.__sb.append("<p>")
        self.__sb.append(string)
        self.__sb.append("</p>\n\n")

    def make_items(self, items: List[str]) -> None:
        self.__sb.append("<ul>")
        for item in items:
            self.__sb.append("<li>")
            self.__sb.append(item)
            self.__sb.append("</li>\n")

        self.__sb.append("\n")

    def close(self) -> None:
        self.__sb.append("</body>")
        self.__sb.append("</html>\n")

        with open(self.__filename, "w") as f:
            f.write(self.__sb.string)

    def get_html_result(self) -> str:
        return self.__filename
